scalameta/tokenizers/shared/src/main/scala/scala/meta/internal/tokenizers/XmlParser.scala

```diff
@@ -132,30 +141,37 @@ class XmlParser(Block: P0, Patterns: P0 = Fail) {
  * Doesn't really parse scala expressions, only reads until the curly brace
  * balance hits 0.
  */
-class ScalaExprPositionParser(dialect: Dialect) extends Parser[Unit] {
+class ScalaExprPositionParser(dialect: Dialect) {
   case class XmlTokenRange(from: Int, to: Int) // from is inclusive, to is exclusive
   private val _splicePositions = List.newBuilder[XmlTokenRange]
   def splicePositions: List[XmlTokenRange] = _splicePositions.result()
 
-  def parseRec(cfg: ParseCtx, index: Int): fastparse.core.Mutable[Unit, Char, String] = {
+  def blockRun = { implicit ctx: ParsingRun[_] =>
     var curlyBraceCount = 1
-    val input = cfg.input
+    val input = ctx.input
+    val index = ctx.index
     val scanner =
       new LegacyScanner(Input.String(input.slice(index, input.length)), dialect)
     scanner.reader.nextChar()
-    while (curlyBraceCount > 0) {
+
+    @tailrec
+    def rec(curlyBraceCount: Int): Boolean = {
       scanner.nextToken()
       (scanner.curr.token: @switch) match {
-        case LegacyToken.LBRACE => curlyBraceCount += 1
-        case LegacyToken.RBRACE => curlyBraceCount -= 1
-        case LegacyToken.EOF =>
-          return fail(cfg.failure, index + scanner.curr.offset)
```

Comment:
so instead of descriptive error, we just return a bool?

Reply:
I think the message will be basically the same via `ctx.freshFailure(nextIndex)`, I am not sure why `cfg.failure` was needed previously, but there is nothing like that in V2